src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ExhaustivenessComputer.java

```diff
@@ -673,5 +782,432 @@ public String toString() {
+     */
+    private Set<PatternDescription> replace(Iterable<? extends PatternDescription> in,
+                                            PatternDescription what,
+                                            Collection<? extends PatternDescription> to) {
+        Set<PatternDescription> result = new HashSet<>();
+
+        for (PatternDescription pd : in) {
+            Collection<? extends PatternDescription> replaced = replace(pd, what, to);
+            if (replaced != null) {
+                result.addAll(replaced);
+            } else {
+                result.add(pd);
+            }
+        }
+
+        return result;
+    }
+    //where:
+        //null: no change
+        private Collection<? extends PatternDescription> replace(PatternDescription in,
+                                                                 PatternDescription what,
+                                                                 Collection<? extends PatternDescription> to) {
+            if (in == what) {
+                return to;
+            } else if (in instanceof RecordPattern rp) {
+                for (int c = 0; c < rp.nested.length; c++) {
+                    Collection<? extends PatternDescription> replaced = replace(rp.nested[c], what, to);
+                    if (replaced != null) {
+                        Set<PatternDescription> withReplaced = new HashSet<>();
+
+                        generatePatternsWithReplacedNestedPattern(rp, c, replaced, Set.of(), withReplaced::add);
+
+                        return replace(withReplaced, what, to);
+                    }
+                }
+                return null;
+            } else {
+                return null; //binding patterns have no children
+            }
+        }
+
+    /* Out of "candidates" remove patterns that are not necessary to achieve exhaustiveness.
+     * Note that iteration order of "candidates" is important - if the set contains
+     * two pattern, out of which either, but not both, is needed to achieve exhaustiveness,
+     * the first one in the iteration order will be removed.
+     */
+    private boolean removeUnnecessaryPatterns(Type selectorType,
+                                              PatternDescription toExpand,
+                                              Set<? extends PatternDescription> basePatterns,
+                                              Set<PatternDescription> inMissingPatterns,
+                                              Set<PatternDescription> candidates) {
+        boolean reduced = false;
+
+        for (Iterator<PatternDescription> it = candidates.iterator(); it.hasNext(); ) {
+            PatternDescription current = it.next();
+            Set<PatternDescription> reducedAdded = new HashSet<>(candidates);
+
+            reducedAdded.remove(current);
+
+            Set<PatternDescription> combinedPatterns =
+                    joinSets(basePatterns, replace(inMissingPatterns, toExpand, reducedAdded));
```

Comment:
```suggestion
    Set<PatternDescription> combinedPatterns =      
              Stream.concat(basePatterns.stream(), 
                                        replace(inMissingPatterns, toExpand, reducedAdded).stream())
                          .collect(Collectors.toSet());
```

Would something like that work, just to avoid the indirection in `joinSets` which is used only once?